arch/arm/arm64/ctx.S

```diff
@@ -0,0 +1,161 @@
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <uk/arch/ctx.h>
+#include <uk/arch/lcpu.h>
+
+/*
+ * NOTE: If we use the SP for single registers, we need to decrease the
+ *       SP 16 bytes for single registers.
+ * https://community.arm.com/arm-community-blogs/b/
+ *  architectures-and-processors-blog/posts/
+ *  using-the-stack-in-aarch64-implementing-push-and-pop
+ *
+ * WARNING: Changes here need also be reflected in uk/asm/ctx.h
+ */
+
+ENTRY(_ctx_arm_clearregs)
+	/* wipe argument and return registers */
+	mov x0, xzr
+	mov x1, xzr
+	mov x2, xzr
+	mov x3, xzr
+	mov x4, xzr
+	mov x5, xzr
+	mov x6, xzr
+	mov x7, xzr
+
+	/* wipe the platform-specific x18 register */
+	mov x18, xzr
+
+	/* wipe callee-saved registers */
+	mov x19, xzr
+	mov x20, xzr
+	mov x21, xzr
+	mov x22, xzr
+	mov x23, xzr
+	mov x24, xzr
+	mov x25, xzr
+	mov x26, xzr
+	mov x27, xzr
+	mov x28, xzr
+#if __OMIT_FRAMEPOINTER__
+	mov x29, xzr /* don't wipe if x29 is used as frame pointer */
+#endif /* __OMIT_FRAMEPOINTER__ */
+
+	/* jump to entrance function left on stack */
+	ldr x30, [sp], #16
+	ret
+
+ENTRY(_ctx_arm_call0)
+#if !__OMIT_FRAMEPOINTER__
+	mov x29, xzr /* reset frame pointer */
+#endif /* !__OMIT_FRAMEPOINTER__ */
+	/* jump to entrance function left on stack */
+	ldr x30, [sp], #16
+	ret
+
+ENTRY(_ctx_arm_call1)
+#if !__OMIT_FRAMEPOINTER__
+	mov x29, xzr /* reset frame pointer */
+#endif /* !__OMIT_FRAMEPOINTER__ */
+	ldr x0, [sp], #16
+	/* jump to entrance function left on stack */
+	ldr x30, [sp], #16
+	ret
+
+ENTRY(_ctx_arm_call2)
+#if !__OMIT_FRAMEPOINTER__
+	mov x29, xzr /* reset frame pointer */
+#endif /* !__OMIT_FRAMEPOINTER__ */
+	ldr x1, [sp], #16
+	ldr x0, [sp], #16
+	/* jump to entrance function left on stack */
+	ldr x30, [sp], #16
+	ret
+
+/*
+ * Do a local context switch taken from thread_start64.S
+ */
+ENTRY(ukarch_ctx_switch)
+
+	/*
+	 * Store the x18 register on the stack.
+	 *
+	 * The role of register x18 is platform specific,
+	 * but saving and restoring it later is definitely
+	 * safer then leaving random values in it.
+         */
+	str x18, [sp, #-16]!
+
+	/* Save callee-saved registers to prevctx's stack */
```

Comment:
```suggestion
	/* Save callee-saved registers to prevctx's stack. */
```

Reply:
I think the dot at the end of one-liner comment is really only personal preference. Since the file has more one-liner comments without ending dot, I'd leave it as is.